plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java

```diff
@@ -168,54 +169,175 @@ protected void canceling() {
+    StagingExitListener stagingExitListener = new StagingExitListener();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, stagingExitListener);
+
+    try {
+      getJobManager().beginRule(project, progress);
+      new ExplodedWarPublisher().publish(project, explodedWarDirectory, progress.newChild(40));
+      new StandardProjectStaging().stage(explodedWarDirectory, stagingDirectory, cloudSdk, progress.newChild(60));
+      return stagingExitListener.getExitStatus();
+    } catch (CoreException | IllegalArgumentException | OperationCanceledException ex) {
+      return StatusUtil.error(getClass(), Messages.getString("deploy.job.staging.failed"), ex);
+    } finally {
+      getJobManager().endRule(project);
+    }
+  }
+
+  private IStatus deployProject(Path credentialFile, IPath stagingDirectory, IProgressMonitor monitor) {
+    DeployExitListener deployExitListener = new DeployExitListener();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, deployExitListener);
+    new AppEngineProjectDeployer().deploy(
+        stagingDirectory, cloudSdk, deployConfiguration, monitor);
+    return deployExitListener.getExitStatus();
   }
 
-  private CloudSdk getCloudSdk(Path credentialFile) {
+  private CloudSdk getCloudSdk(Path credentialFile, ProcessExitListener processExitListener) {
     CloudSdk cloudSdk = new CloudSdk.Builder()
                           .addStdOutLineListener(stdoutLineListener)
                           .addStdErrLineListener(stderrLineListener)
                           .addStdErrLineListener(errorCollectingLineListener)
                           .appCommandCredentialFile(credentialFile.toFile())
                           .startListener(new StoreProcessObjectListener())
-                          .exitListener(new RecordProcessError())
+                          .exitListener(processExitListener)
                           .appCommandMetricsEnvironment(CloudToolsInfo.METRICS_NAME)
                           .appCommandMetricsEnvironmentVersion(CloudToolsInfo.getToolsVersion())
+                          .appCommandOutputFormat("json")
                           .build();
     return cloudSdk;
   }
 
+  private IStatus openAppInBrowser() {
+    final String project = deployConfiguration.getProject();
+    String appLocation = null;
+    if (deployConfiguration.getPromote()) {
+      appLocation = "http://" + project + ".appspot.com";
+    } else {
+      try {
+        String version =  getDeployedAppVersion();
+        appLocation = "http://" + version + "-dot-" + project+ ".appspot.com/";
+      } catch (IndexOutOfBoundsException | JsonParseException ex)  {
+        return StatusUtil.error(getClass(), Messages.getString("brower.launch.failed"), ex);
+      }
+    }
+
+    final String finalAppLocation = appLocation;
+    final IWorkbench workbench = PlatformUI.getWorkbench();
+    Job launchBrowserJob = new UIJob(workbench.getDisplay(), "Launch deployed app in browser") {
```

Comment:
This would make a good utility class, and we have a few locations that would likely benefit from it too.